File: ios/TransportHQ/Models/Trip.swift
import Foundation

struct Trip: Codable, Identifiable {
    let id: String
    let rideRequest: TripRideRequest?
    let driver: TripDriver?
    let vehicle: TripVehicle?
    let status: TripStatus
    let pickupLocation: TripLocation?
    let dropoffLocation: TripLocation?
    let passengers: [TripPassengerStop]?
    let scheduledTime: String?
    let startTime: String?
    let endTime: String?
    let eta: String?
    let currentLocation: GeoPoint?
    let distance: Double?
    let duration: Double?
    let createdAt: String?
    let updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case rideRequest, driver, vehicle, status
        case pickupLocation, dropoffLocation, passengers
        case scheduledTime, startTime, endTime, eta
        case currentLocation, distance, duration
        case createdAt, updatedAt
    }

    var displayPickup: String {
        pickupLocation?.address
            ?? rideRequest?.pickupLocation?.address
            ?? passengers?.first?.pickupAddress
            ?? "N/A"
    }

    var displayDropoff: String {
        dropoffLocation?.address
            ?? rideRequest?.dropoffLocation?.address
            ?? passengers?.first?.dropoffAddress
            ?? "N/A"
    }
}

struct TripPassengerStop: Codable {
    let id: String?
    let name: String?
    let phone: String?
    let pickupAddress: String?
    let dropoffAddress: String?
    let pickupTime: String?
    let dropoffTime: String?
    let status: String?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case name, phone, pickupAddress, dropoffAddress
        case pickupTime, dropoffTime, status
    }
}

struct TripRideRequest: Codable {
    let id: String?
    let passenger: TripPassenger?
    let pickupLocation: TripLocation?
    let dropoffLocation: TripLocation?
    let priority: String?
    let department: String?
    let notes: String?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case passenger, pickupLocation, dropoffLocation, priority, department, notes
    }
}

struct TripPassenger: Codable {
    let id: String?
    let name: String?
    let firstName: String?
    let lastName: String?
    let phone: String?
    let email: String?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case name, firstName, lastName, phone, email
    }

    var fullName: String {
        if let name = name, !name.isEmpty { return name }
        return "\(firstName ?? "") \(lastName ?? "")".trimmingCharacters(in: .whitespaces)
    }
}

struct TripDriver: Codable {
    let id: String?
    let name: String?
    let firstName: String?
    let lastName: String?
    let phone: String?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case name, firstName, lastName, phone
    }

    var fullName: String {
        if let name = name, !name.isEmpty { return name }
        return "\(firstName ?? "") \(lastName ?? "")".trimmingCharacters(in: .whitespaces)
    }
}

struct TripVehicle: Codable {
    let id: String?
    let make: String?
    let model: String?
    let year: Int?
    let licensePlate: String?
    let type: String?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case make, model, year, licensePlate, type
    }

    var displayName: String {
        "\(make ?? "") \(model ?? "") (\(licensePlate ?? ""))".trimmingCharacters(in: .whitespaces)
    }
}

struct TripLocation: Codable {
    let address: String?
    let coordinates: GeoPoint?
}

struct GeoPoint: Codable {
    let lat: Double?
    let lng: Double?

    enum CodingKeys: String, CodingKey {
        case lat, lng
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        lat = try? container.decodeIfPresent(Double.self, forKey: .lat)
        lng = try? container.decodeIfPresent(Double.self, forKey: .lng)
    }

    init(lat: Double, lng: Double) {
        self.lat = lat
        self.lng = lng
    }
}

enum TripStatus: String, Codable {
    case unassigned
    case assigned
    case driverDeparted = "driver-departed"
    case arrivedPickup = "arrived-pickup"
    case inProgress = "in-progress"
    case completed
    case cancelled

    var displayName: String {
        switch self {
        case .unassigned: return "Unassigned"
        case .assigned: return "Assigned"
        case .driverDeparted: return "Driver Departed"
        case .arrivedPickup: return "Arrived at Pickup"
        case .inProgress: return "In Progress"
        case .completed: return "Completed"
        case .cancelled: return "Cancelled"
        }
    }

    var nextStatus: TripStatus? {
        switch self {
        case .unassigned: return nil
        case .assigned: return .driverDeparted
        case .driverDeparted: return .arrivedPickup
        case .arrivedPickup: return .inProgress
        case .inProgress: return .completed
        case .completed, .cancelled: return nil
        }
    }

    var nextStatusLabel: String? {
        switch self {
        case .unassigned: return nil
        case .assigned: return "I'm Departing Now"
        case .driverDeparted: return "I've Arrived at Pickup"
        case .arrivedPickup: return "Passenger Picked Up"
        case .inProgress: return "Arrived at Destination"
        case .completed, .cancelled: return nil
        }
    }

    var isActive: Bool {
        switch self {
        case .driverDeparted, .arrivedPickup, .inProgress:
            return true
        default:
            return false
        }
    }
}

struct TripDetailResponse: Codable {
    let trip: Trip
    let events: [TripEvent]?
}

struct TripResponse: Codable {
    let trip: Trip
}

struct TripEvent: Codable, Identifiable {
    let id: String
    let trip: String?
    let type: String?
    let actor: TripEventActor?
    let details: String?
    let timestamp: String?
    let createdAt: String?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case trip, type, actor, details, timestamp, createdAt
    }
}

struct TripEventActor: Codable {
    let id: String?
    let name: String?
    let role: String?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case name, role
    }
}

File: android/app/src/main/java/com/transporthq/app/data/models/Trip.kt
package com.transporthq.app.data.models

import com.google.gson.annotations.SerializedName

data class Trip(
    @SerializedName("_id") val id: String = "",
    val rideRequest: RideRequestRef? = null,
    val driver: DriverRef? = null,
    val vehicle: VehicleRef? = null,
    val status: String = "assigned",
    val pickupLocation: String? = null,
    val dropoffLocation: String? = null,
    val passengers: List<TripPassengerStop>? = null,
    val scheduledTime: String? = null,
    val startTime: String? = null,
    val endTime: String? = null,
    val passengerCount: Int = 1,
    val notes: String? = null,
    val createdAt: String? = null,
    val updatedAt: String? = null
) {
    val displayPickup: String
        get() = pickupLocation
            ?: rideRequest?.pickupLocation
            ?: passengers?.firstOrNull()?.pickupAddress
            ?: "N/A"

    val displayDropoff: String
        get() = dropoffLocation
            ?: rideRequest?.dropoffLocation
            ?: passengers?.firstOrNull()?.dropoffAddress
            ?: "N/A"
}

data class TripPassengerStop(
    @SerializedName("_id") val id: String = "",
    val name: String = "",
    val phone: String = "",
    val pickupAddress: String = "",
    val dropoffAddress: String = "",
    val pickupTime: String? = null,
    val dropoffTime: String? = null,
    val status: String = "waiting",
    val user: PassengerRef? = null
)

data class RideRequestRef(
    @SerializedName("_id") val id: String = "",
    val pickupLocation: String? = null,
    val dropoffLocation: String? = null,
    val passenger: PassengerRef? = null,
    val requester: PassengerRef? = null,
    val passengerCount: Int = 1,
    val priority: String = "normal",
    val notes: String? = null
)

data class DriverRef(
    @SerializedName("_id") val id: String = "",
    val name: String = "",
    val phone: String = "",
    val email: String = "",
    val status: String = "available"
)

data class VehicleRef(
    @SerializedName("_id") val id: String = "",
    val name: String = "",
    @SerializedName("licensePlate") val plateNumber: String = "",
    val type: String = "",
    val capacity: Int = 4
)

data class PassengerRef(
    @SerializedName("_id") val id: String = "",
    val name: String = "",
    val email: String = "",
    val phone: String = "",
    val department: String = ""
)

data class LocationInfo(
    val address: String = "",
    val coordinates: Coordinates? = null
)

data class Coordinates(
    val lat: Double = 0.0,
    val lng: Double = 0.0
)

data class TripStatusUpdate(
    val status: String,
    val location: Coordinates? = null
)

data class TripsResponse(
    val trips: List<Trip> = emptyList()
)

data class TripDetailResponse(
    val trip: Trip,
    val events: List<TripEvent> = emptyList()
)

data class TripResponse(
    val trip: Trip
)

data class TripEvent(
    @SerializedName("_id") val id: String = "",
    val trip: String = "",
    val type: String = "",
    val actor: TripEventActor? = null,
    val details: String? = null,
    val location: Coordinates? = null,
    val timestamp: String? = null,
    val createdAt: String? = null
)

data class TripEventActor(
    @SerializedName("_id") val id: String = "",
    val name: String = "",
    val role: String = ""
)

data class CreateTripRequest(
    val title: String,
    val date: String,
    val driverId: String? = null,
    val vehicleId: String? = null,
    val passengers: List<TripPassengerInput> = emptyList(),
    val pickupLocation: String = "",
    val dropoffLocation: String = "",
    val scheduledTime: String? = null,
    val notes: String = ""
)

data class TripPassengerInput(
    val userId: String,
    val name: String = ""
)

data class AssignTripRequest(
    val driverId: String,
    val vehicleId: String
)
